src/mcp/client/auth.py

```diff
@@ -510,9 +513,28 @@ async def async_auth_flow(self, request: httpx.Request) -> AsyncGenerator[httpx.
                 try:
                     # OAuth flow must be inline due to generator constraints
                     # Step 1: Discover protected resource metadata (RFC9728 with WWW-Authenticate support)
-                    discovery_request = await self._discover_protected_resource(response)
-                    discovery_response = yield discovery_request
-                    await self._handle_protected_resource_response(discovery_response)
+                    # Check if WWW-Authenticate provides resource_metadata URL first
+                    www_auth_url = self._extract_resource_metadata_from_www_auth(response)
+                    if www_auth_url:
+                        discovery_request = httpx.Request(
+                            "GET", www_auth_url, headers={MCP_PROTOCOL_VERSION: LATEST_PROTOCOL_VERSION}
+                        )
+                        discovery_response = yield discovery_request
+                        await self._handle_protected_resource_response(discovery_response)
+                    else:
+                        # Try well-known discovery URLs with fallback
+                        discovery_urls = self._get_protected_resource_discovery_urls()
+                        for url in discovery_urls:
+                            discovery_request = httpx.Request(
+                                "GET", url, headers={MCP_PROTOCOL_VERSION: LATEST_PROTOCOL_VERSION}
+                            )
+                            discovery_response = yield discovery_request
+
+                            if discovery_response.status_code == 200:
+                                await self._handle_protected_resource_response(discovery_response)
+                                break  # Success, stop trying other URLs
+                            elif discovery_response.status_code != 404:
+                                break  # Non-404 error, stop trying
```

Comment:
In practice we may want to keep retrying on other types of errors too, just to be safe